pkg/deploy/job.go

```diff
@@ -22,58 +26,113 @@ import (
 	"github.com/sirupsen/logrus"
 	batchv1 "k8s.io/api/batch/v1"
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
-var jobDiffOpts = cmp.Options{
-	cmpopts.IgnoreFields(batchv1.Job{}, "TypeMeta", "ObjectMeta", "Status"),
+const (
+	CheTlsJobComponentName = "che-create-tls-secret-job"
+)
+
+var (
+	jobDiffOpts = cmp.Options{
+		cmpopts.IgnoreFields(batchv1.Job{}, "TypeMeta", "ObjectMeta", "Status"),
+		cmpopts.IgnoreFields(batchv1.JobSpec{}, "Selector", "TTLSecondsAfterFinished"),
+		cmpopts.IgnoreFields(v1.PodTemplateSpec{}, "ObjectMeta"),
+		cmpopts.IgnoreFields(corev1.Container{}, "TerminationMessagePath", "TerminationMessagePolicy"),
+		cmpopts.IgnoreFields(corev1.PodSpec{}, "DNSPolicy", "SchedulerName", "SecurityContext"),
+		cmp.Comparer(func(x, y []corev1.EnvVar) bool {
+			xMap := make(map[string]string)
+			yMap := make(map[string]string)
+			for _, env := range x {
+				xMap[env.Name] = env.Value
+			}
+			for _, env := range y {
+				yMap[env.Name] = env.Value
+			}
+			return reflect.DeepEqual(xMap, yMap)
+		}),
+	}
+)
+
+func SyncTlsJobToCluster(
+	checluster *orgv1.CheCluster,
+	name string,
+	image string,
+	serviceAccountName string,
+	env map[string]string,
+	clusterAPI ClusterAPI) (*batchv1.Job, reconcile.Result, error) {
+	return SyncJobToCluster(checluster, name, CheTlsJobComponentName, image, serviceAccountName, env, clusterAPI)
 }
 
-// SyncJobToCluster deploys new instance of given job
-func SyncJobToCluster(instance *orgv1.CheCluster, job *batchv1.Job, clusterAPI ClusterAPI) error {
-	jobFound, err := GetClusterJob(job.ObjectMeta.Name, job.ObjectMeta.Namespace, clusterAPI)
-	if jobFound == nil && err == nil {
-		logrus.Infof("Creating a new object: %s, name: %s", job.Kind, job.Name)
-		err = clusterAPI.Client.Create(context.TODO(), job)
+func SyncJobToCluster(
+	checluster *orgv1.CheCluster,
+	name string,
+	component string,
+	image string,
+	serviceAccountName string,
+	env map[string]string,
+	clusterAPI ClusterAPI) (*batchv1.Job, reconcile.Result, error) {
```

Comment:
Isn't it a general helper? Please move reconcile decision into controller package.